lib/puppet/pops/types/p_object_type.rb

```diff
@@ -0,0 +1,539 @@
+      super(name, TypeAsserter.assert_instance_of(["initializer for function '%s'", name], TYPE_FUNCTION, i12n_hash))
+    end
+
+    # @api private
+    def feature_type
+      'function'
+    end
+  end
+
+  attr_reader :parent
+  attr_reader :attributes
+  attr_reader :functions
+  attr_reader :equality
+  attr_reader :checks
+  attr_reader :annotations
+
+  # @param i12n_hash [Hash{String=>Object}] The hash describing the Object features
+  #
+  # @api public
+  def initialize(i12n_hash)
+    TypeAsserter.assert_instance_of('object initializer', TYPE_OBJECT_I12N, i12n_hash)
+
+    @attributes = EMPTY_HASH
+    @functions = EMPTY_HASH
+    @parent = i12n_hash[KEY_PARENT]
+
+    parent_members = EMPTY_HASH
+    parent_object_type = nil
+    unless @parent.nil?
+      rp = resolved_parent
+      if rp.is_a?(PObjectType)
+        parent_object_type = rp
+        parent_members = rp.members(true)
+      end
+    end
+
+    attr_specs = i12n_hash[KEY_ATTRIBUTES]
+    unless attr_specs.nil? || attr_specs.empty?
+      @attributes = Hash[attr_specs.map do |key, attr_spec|
+        attr_spec = { KEY_TYPE => TypeAsserter.assert_instance_of(["attribute '%s'", key], PType::DEFAULT, attr_spec) } unless attr_spec.is_a?(Hash)
+        attr = PAttribute.new(key, attr_spec)
+        [attr.name, attr.assert_override(parent_members)]
+      end].freeze
+    end
+
+    func_specs = i12n_hash[KEY_FUNCTIONS]
+    unless func_specs.nil? || func_specs.empty?
+      @functions = Hash[func_specs.map do |key, func_spec|
+        func_spec = { KEY_TYPE => TypeAsserter.assert_instance_of(["function '%s'", key], TYPE_FUNCTION_TYPE, func_spec) } unless func_spec.is_a?(Hash)
+        func = PFunction.new(key, func_spec)
+        name = func.name
+        raise Puppet::ParseError, "function '#{key}' conflicts with attribute with the same name" if @attributes.include?(name)
+        [name, func.assert_override(parent_members)]
+      end].freeze
+    end
+
+    @equality_include_type = i12n_hash[KEY_EQUALITY_INCLUDE_TYPE]
+    @equality_include_type = true if @equality_include_type.nil?
+
+    equality = i12n_hash[KEY_EQUALITY]
+    equality = [equality] if equality.is_a?(String)
+    if equality.is_a?(Array)
+      unless equality.empty?
+        raise Puppet::ParseError, 'equality_include_type = false cannot be combined with attributes' unless @equality_include_type
+        parent_eq_attrs = nil
+        equality.each do |attr_name|
+
+          attr = parent_members[attr_name]
+          if attr.nil?
+            attr = @attributes[attr_name] || @functions[attr_name]
+          elsif attr.is_a?(PAttribute)
+            # Assert that attribute is not already include by parent equality
+            parent_eq_attrs ||= parent_object_type.equality_attributes
+            if parent_eq_attrs.include?(attr_name)
+              raise Puppet::ParseError, "equality is referencing attribute '#{attr_name}' which is already included by parent equality"
```

Comment:
There are two questions that might need an answer.

1. In which parent was the attribute defined?
2. In which parent was the attribute added to the equality definition?

Just looking where the actual attribute will not be enough since it might be added by a sub-type to the type where the attribute is declared. In order to figure this out, it will be necessary to traverse the parent chain and check which of the parents it is that actually adds the attribute. Is the added info worth that kind of complexity?